src/main/java/net/dries007/tfc/common/entities/ai/TFCGroundPathNavigation.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Licensed under the EUPL, Version 1.2.
+ * You may obtain a copy of the Licence at:
+ * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
+ */
+
+package net.dries007.tfc.common.entities.ai;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.tags.FluidTags;
+import net.minecraft.util.Mth;
+import net.minecraft.world.entity.Mob;
+import net.minecraft.world.entity.ai.navigation.GroundPathNavigation;
+import net.minecraft.world.level.BlockGetter;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.pathfinder.*;
+import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.Vec3;
+
+import net.dries007.tfc.util.Helpers;
+
+public class TFCGroundPathNavigation extends GroundPathNavigation
+{
+    public static int getSurfaceYRespectingFluid(GroundPathNavigation navigation, Mob mob, Level level)
+    {
+        final BlockPos.MutableBlockPos cursor = new BlockPos.MutableBlockPos();
+        if (mob.isInWater() && navigation.canFloat())
+        {
+            BlockState state = level.getBlockState(cursor.set(mob.getX(), mob.getBlockY(), mob.getZ()));
+            int checked = 0;
+
+            while (Helpers.isFluid(state.getFluidState(), FluidTags.WATER))
+            {
+                state = level.getBlockState(cursor.move(0, 1, 0));
+                ++checked;
+                if (checked > 16)
+                {
+                    return mob.getBlockY();
+                }
+            }
+            return cursor.getY();
+        }
+        return Mth.floor(mob.getY() + 0.5D);
+    }
+
+    public TFCGroundPathNavigation(Mob mob, Level level)
+    {
+        super(mob, level);
+    }
+
+    @Override
+    protected PathFinder createPathFinder(int followRange)
+    {
+        nodeEvaluator = new TFCWalkNodeEvaluator();
+        nodeEvaluator.setCanPassDoors(true);
+        return new PathFinder(nodeEvaluator, followRange);
+    }
+
+    @Override
+    protected Vec3 getTempMobPos()
+    {
+        return new Vec3(mob.getX(), getSurfaceYRespectingFluid(this, mob, level), mob.getZ());
+    }
+
+    public static class TFCWalkNodeEvaluator extends WalkNodeEvaluator
+    {
+        @Override
+        protected BlockPathTypes evaluateBlockPathType(BlockGetter level, boolean openDoors, boolean enterDoors, BlockPos pos, BlockPathTypes pathType)
+        {
+            // interpret leaves as open space
```

Comment:
Should we be setting this in the leaves block? i.e. the `getBlockPathType` method or whatnot? It seems weird to be special casing it here.

Reply:
generally we want some things to recognize leaves as leaves (aka something blocked), for example if another mod added a bird mob that lands on trees.

Reply:
in vanilla leaves are solid tho; and ravagers straight out break leaves. I don't think that's comparable, no?